src/cargo/core/compiler/mod.rs

```diff
@@ -629,19 +631,42 @@ fn rustdoc<'a, 'cfg>(cx: &mut Context<'a, 'cfg>, unit: &Unit<'a>) -> CargoResult
     }))
 }
 
-fn add_crate_versions_if_requested(
-    bcx: &BuildContext<'_, '_>,
-    unit: &Unit<'_>,
+fn add_crate_versions_if_requested<'a>(
+    bcx: &BuildContext<'a, '_>,
+    unit: &Unit<'a>,
     rustdoc: &mut ProcessBuilder,
 ) {
-    if !bcx.config.cli_unstable().crate_versions {
-        return;
+    if crate_versions_requested(bcx)
+        && !crate_version_flag_in_rustdocflags(bcx, unit)
+        && !crate_version_flag_in_extra_args(bcx, unit)
```

Comment:
I'm thinking it might be a little simpler just to check the args of the `ProcessBuilder` instead of re-fetching the extra args/rustdocflags.

So, move the call to `add_crate_versions_if_requested` after the args are added to the builder, and call something like this:

```rust
rustdoc.get_args().iter().any(|flag| {
    flag.to_str().map_or(false, |s| s.starts_with(RUSTDOC_CRATE_VERSION_FLAG))
})
```